src/FSharp.Core/list.fs

```diff
@@ -987,3 +988,173 @@ module List =
 
         coll.AddMany(values) // insert values BEFORE the item at the index
         coll.AddManyAndClose(curr)
+
+    [<CompiledName("RandomShuffleWith")>]
+    let randomShuffleWith (random: Random) (source: 'T list) : 'T list =
+        checkNonNull "random" random
+
+        let tempArray = toArray source
+        Microsoft.FSharp.Primitives.Basics.Random.shuffleArrayInPlaceWith random tempArray
+        ofArray tempArray
+
+    [<CompiledName("RandomShuffleBy")>]
+    let randomShuffleBy (randomizer: unit -> float) (source: 'T list) : 'T list =
+        let tempArray = toArray source
+        Microsoft.FSharp.Primitives.Basics.Random.shuffleArrayInPlaceBy randomizer tempArray
+        ofArray tempArray
+
+    [<CompiledName("RandomShuffle")>]
+    let randomShuffle (source: 'T list) : 'T list =
+        randomShuffleWith ThreadSafeRandom.Shared source
+
+    [<CompiledName("RandomChoiceWith")>]
+    let randomChoiceWith (random: Random) (source: 'T list) : 'T =
+        checkNonNull "random" random
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        let i = random.Next(0, inputLength)
+        source[i]
+
+    [<CompiledName("RandomChoiceBy")>]
+    let randomChoiceBy (randomizer: unit -> float) (source: 'T list) : 'T =
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        let i = Microsoft.FSharp.Primitives.Basics.Random.next randomizer 0 inputLength
+        source[i]
+
+    [<CompiledName("RandomChoice")>]
+    let randomChoice (source: 'T list) : 'T =
+        randomChoiceWith ThreadSafeRandom.Shared source
+
+    [<CompiledName("RandomChoicesWith")>]
+    let randomChoicesWith (random: Random) (count: int) (source: 'T list) : 'T list =
+        checkNonNull "random" random
+
+        if count < 0 then
+            invalidArgInputMustBeNonNegative "count" count
+
+        let inputLength = source.Length
+
+        if inputLength = 0 then
+            invalidArg "source" LanguagePrimitives.ErrorStrings.InputSequenceEmptyString
+
+        [
+            for _ = 0 to count - 1 do
+                let j = random.Next(0, inputLength)
+                source[j]
+        ]
+
+    [<CompiledName("RandomChoicesBy")>]
+    let randomChoicesBy (randomizer: unit -> float) (count: int) (source: 'T list) : 'T list =
```

Comment:
For sampling, what stops us from reusing the array implementation here and in sequences? (like it's done with shuffle for example)

Apart from the empty collection check, I see some differences in the element copy. Do I get it right that it leads to optimal performance of these operations within different collections?

Asking just for my understanding, I believe these algorithms are not going to change quickly so some code duplication is not the end of the world here.

Reply:
It's just that there's no easy way to reuse the array implementation :)
1) For array we have `zeroCreateUnchecked` which is the fastest way to create a collection and it happens that this is exactly the type we need
2) For list there is no point in creating extra array, since we can initiate list just from the input list
3) Sequences are the same as list, but we need to copy all the sequence to array first to reduce amount of iterations. The difference with list is that we don't have to call randomizer immediately for all elements - we are returning a sequence that will be lazily invoke randomizer as requested. For shuffling the reuse is possible, because the temp array matches the required argument.

Reply:
Right, yeah that makes sense, thanks for the explanation!